### PR TITLE
dwarf: extend support for DWARF formats

### PR DESCRIPTION
**Pull request by MarekPazur**
DWARF debug information may be stored in various external formats, including:
 - stripped `.debug` files
 - split DWARF object files and packages (`.dwo`, `.dwp`)
 - deduplicated shared `.dwz` files

In some cases, it is also necessary to have the external debug files outside the standard search paths, e.g. when running inside a container.

This commit enables the parser to handle these formats by adding support for dwo/dwp parsing and introducing a `debuginfo` option to extend the debug info search paths.

The `debuginfo` option does not yet support dwo/dwp files due to limitations in the libdw API. Support via debuginfod is currently under development in elfutils (as of 2026).

Additionally, fix a bug where a missing .dwz file could cause a crash by checking for file existence before parsing. This is not covered in tests on purpose, as the DWZ tool is very unstable and therefore difficult to test properly.

##### Checklist

- [x] Language changes are updated in `docs/language.md`, `docs/stdlib.md`, or `man/adoc/bpftrace.adoc`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

### CODE REVIEW

File: man/adoc/bpftrace.adoc [hunk 1]
```diff
@@ -80,6 +80,18 @@ For example, you could add the predicate `/pid == cpid/` to probes with userspac
 Enable debug mode.
 For more details see the <<Debug Output>> section.
 
+=== *--debuginfo-path*
```
Comment (jordalgo) [inline man/adoc/bpftrace.adoc]:
```suggestion
=== *--debuginfo-path* DIR
```

File: man/adoc/bpftrace.adoc [hunk 2]
```diff
@@ -80,6 +80,18 @@ For example, you could add the predicate `/pid == cpid/` to probes with userspac
...
+
+Add the directory DIR to the search path for DWARF debug information.
+Paths may be absolute or relative to the traced binary's location.
+Multiple paths can be specified by separating them with a colon (`:`), for example `--debuginfo-path=/bin/debug:./debug:..`.
+Files found in these paths are validated using the build ID.
+Alternatively, CRC32 checksum validation can be enabled by prefixing a path with `+`.
```
Comment (jordalgo) [inline man/adoc/bpftrace.adoc:89]: Is this a standard thing?

File: man/adoc/bpftrace.adoc [hunk 3]
```diff
@@ -80,6 +80,18 @@ For example, you could add the predicate `/pid == cpid/` to probes with userspac
...
+Files that do not match are skipped.
+By default, bpftrace searches standard debug paths or attempts to query available debuginfod servers.
+
+At the moment, split debuginfo format (`.dwo`, `.dwp`) is not supported for this option.
```
Comment (jordalgo) [inline man/adoc/bpftrace.adoc]: nit:
```suggestion
Split debuginfo format (`.dwo`, `.dwp`) is not yet supported for this option.
```

File: src/main.cpp [hunk 4]
```diff
@@ -149,6 +150,8 @@ void usage(std::ostream& out)
   out << "                   only run probes whose name matches REGEX" << std::endl;
   out << "    --traceable-functions FILE" << std::endl;
   out << "                   load the list of traceable kernel functions from FILE" << std::endl;
+  out << "    --debuginfo-path DIR" << std::endl;
+  out << "                   add the directory to the debug info search path" << std::endl;
```
Comment (jordalgo) [inline src/main.cpp]:
```suggestion
  out << "                   add one or more directory to the debug info search path" << std::endl;
```

File: man/adoc/bpftrace.adoc (hunk 3 above)
Comment (jordalgo) [reply man/adoc/bpftrace.adoc]: Also, maybe we should add a note that split debuginfo works for the normal probe binary path.

**Review by jordalgo — COMMENTED** [review]
Some AI review comments:
1. src/dwarf_parser.cpp:31 / src/dwarf_parser.h:120: the new debuginfo_path_ member stores debuginfo_path.c_str() from a caller-owned std::string,
     and that borrowed pointer is then exposed to libdwfl for the lifetime of the Dwarf object. GetFromBinary() now accepts any const std::string &,
     so a temporary or later-mutated source string leaves callbacks.debuginfo_path dangling and can turn debuginfo lookup into a use-after-free. This
     should be owned as a std::string inside Dwarf, with the callback pointing at that storage.
  2. src/dwarf_parser.cpp:108: the new warning path for missing split debuginfo unconditionally does std::filesystem::path(dwo_name) even though
     dwo_name is initialized to nullptr and only populated if DW_AT_dwo_name/DW_AT_GNU_dwo_name is present and string-formatted successfully. For a
     skeleton CU without that attribute, or with an unreadable form, this warning path can crash instead of degrading gracefully.

File: src/dwarf_parser.h [hunk 5]
```diff
@@ -58,13 +60,34 @@ class Dwarf {
...
+    // manually per the libdw API, and is valid for the lifetime of the CuInfo
+    // instance.
+    //
+    // Present when cudie is a DW_UT_skeleton unit. In that case, this should
+    // be used instead, as it holds vast majority of the debug info. libdw
+    // automatically resolves references between the skeleton and split CU.
+    std::optional<Dwarf_Die> split_cudie;
+
+    // Returns split CU DIE if present (skeleton CU), otherwise cudie.
+    Dwarf_Die *cu_die()
```
Comment (jordalgo) [inline src/dwarf_parser.h:86]:
```suggestion
    const Dwarf_Die *cu_die()
```

File: man/adoc/bpftrace.adoc (hunk 2 above)
Comment (MarekPazur) [reply man/adoc/bpftrace.adoc:89]: Its a part of the elfutils libdwfl behaviour, so I exposed it for consistency reasons, although it specifically isn't a convention. Should I remove this text from the docs?

> If the first character of the string is + or - that enables or disables CRC32 checksum validation when it's necessary.  The remainder of the string is composed of elements separated by colons.  Each element can start with + or - to override the global checksum behavior.  This flag is never relevant when working with build IDs, but it's always parsed in the path string. Any file found in the path is validated by build ID if possible, or else by CRC32 checksum if enabled, and skipped if it does not match.

File: src/dwarf_parser.h (hunk 5 above)
Comment (MarekPazur) [reply src/dwarf_parser.h:86]: The library API doesn't take const parameters, so every function call would need an immediate `const_cast` anyways. Do you insist on making it const?

File: man/adoc/bpftrace.adoc (hunk 2 above)
Comment (jordalgo) [reply man/adoc/bpftrace.adoc:89]: I think it's fine to leave it if it's supported and users might want to use it. I was mostly just curious.

File: src/dwarf_parser.h (hunk 5 above)
Comment (jordalgo) [reply src/dwarf_parser.h:86]: Nah, ignore my comment if it doesn't work.

**Comment by MarekPazur** [comment]
Should be done now.

File: man/adoc/bpftrace.adoc [hunk 6]
```diff
@@ -80,6 +80,18 @@ For example, you could add the predicate `/pid == cpid/` to probes with userspac
 Enable debug mode.
 For more details see the <<Debug Output>> section.
 
+=== *--debuginfo-path* _DIR_
```
Comment (jordalgo) [inline man/adoc/bpftrace.adoc]: Super nit - what do you think about just shortening this to `--debuginfo` ? It's not as explicit but, on the other hand, I'm not sure what else we might use a `--debuginfo` flag for.

File: src/dwarf_parser.cpp [hunk 7]
```diff
@@ -27,25 +28,50 @@ struct FuncInfo {
...
 
-Dwarf::Dwarf(BPFtrace *bpftrace, const std::string &file_path)
-    : bpftrace_(bpftrace), file_path_(file_path)
+Dwarf::Dwarf(BPFtrace *bpftrace,
+             const std::string &file_path,
+             std::string debuginfo_path)
+    : bpftrace_(bpftrace),
+      file_path_(file_path),
+      debuginfo_path_(std::move(debuginfo_path)),
+      debuginfo_path_cstr_(debuginfo_path_.c_str())
```
Comment (jordalgo) [inline src/dwarf_parser.cpp]: Let's also delete the copy/move constructors for this class just to make sure. Also AI complains about setting the cstr_ in the initializer list and says it's better practice to do this in the constructor body. 🤷

File: src/dwarf_parser.cpp [hunk 8]
```diff
@@ -427,84 +500,74 @@ ssize_t Dwarf::get_bitfield_size(Dwarf_Die &field_die)
...
+  if (!src_path) {
+    return make_error<DwarfParseError>(
+        "Failed to get compilation unit source path");
+  }
+
   Dwarf_Lines *lines = nullptr;
   size_t num_lines = 0;
 
-  if (dwarf_getsrclines(cu->die, &lines, &num_lines) != 0) {
+  if (dwarf_getsrclines(cu->cudie, &lines, &num_lines) != 0) {
```
Comment (jordalgo) [inline src/dwarf_parser.cpp]: From claude: 
  src/dwarf_parser.cpp:577 — dwarf_getsrclines(cu->cudie, ...) accesses the raw cudie field directly instead of using cu->cu_die(). For skeleton CUs
  with split DWARF, cudie points to the skeleton, which may not contain source line info. This should likely be cu->cu_die() for consistency, unless
  dwarf_getsrclines specifically requires the skeleton CU (which would be worth a comment).

  This looks like a bug — if a user has split DWARF and tries to use source line attach points, dwarf_getsrclines on the skeleton CU may fail or
  return incomplete results.

File: tests/testprogs/CMakeLists.txt [hunk 9]
```diff
@@ -96,6 +96,51 @@ add_custom_command(
...
+  ${CMAKE_CURRENT_BINARY_DIR}/debug/dwp
+)
+
+# Create a stripped binary with a separate .debug file, placed in a subdir; used to verify
+# loading and parsing of external debuginfo.
+add_custom_command(
+  OUTPUT 
+    ${CMAKE_CURRENT_BINARY_DIR}/uprobe_test-stripped
+    ${CMAKE_CURRENT_BINARY_DIR}/debug/uprobe_test-stripped.debug
+  COMMAND ${LLVM_OBJCOPY} --only-keep-debug $<TARGET_FILE:uprobe_test> ${CMAKE_CURRENT_BINARY_DIR}/debug/uprobe_test-stripped.debug
```
Comment (jordalgo) [inline tests/testprogs/CMakeLists.txt:112]: We should also add `--add-gnu-debuglink`

**Review by jordalgo — COMMENTED** [review]
One more AI comment - plz let me know if these are BS:

src/dwarf_parser.cpp:135 mixes dwarf_hasattr_integrate() with dwarf_diename() on the original DIE. For split/inlined subprogram DIEs whose DW_AT_name is only available through DW_AT_specification or DW_AT_abstract_origin, the integrated attribute check can succeed while dwarf_diename(func_die) is still null. That makes the comparison at src/dwarf_parser.cpp:136 either miss valid functions or hit undefined behavior when the C string is compared to std::string. The name needs to be read from the integrated attribute itself, not from dwarf_diename() on the raw DIE.

File: man/adoc/bpftrace.adoc (hunk 6 above)
Comment (MarekPazur) [reply man/adoc/bpftrace.adoc]: Sure, sounds good.

File: src/dwarf_parser.cpp (hunk 8 above)
Comment (MarekPazur) [reply src/dwarf_parser.cpp]: This is on purpose, as the line info is present in the skeleton CUs at all times. `cu->cu_die()` would work aswell, as it is linked with the split CU, it just adds a small overhead. But that may be negligible anyways, so I'll change it to be consistent with the rest of this file.

**Comment by MarekPazur** [comment]
This should be relatively okay-ish, as both `dwarf_hasattr_integrate()` and `dwarf_diename()` check the presence of the attribute. Also `dwarf_diename()` does read the name (data) from the attribute itself, as it internally uses `dwarf_attr_integrate()`. However it can return `nullptr`, when the data is seriously malformed, which should be difficult to achieve, still possible though.

In this file, `dwarf_diename()` is used a lot in the same way, so maybe a wrapper could make it safer.

```
std::string_view Dwarf::get_diename(Dwarf_Die *die)
{
    const char* name = dwarf_diename(die);
    return name ? name : "";
}
```

> src/dwarf_parser.cpp:135 mixes dwarf_hasattr_integrate() with dwarf_diename() on the original DIE. For split/inlined subprogram DIEs whose DW_AT_name is only available through DW_AT_specification or DW_AT_abstract_origin, the integrated attribute check can succeed while dwarf_diename(func_die) is still null. That makes the comparison at src/dwarf_parser.cpp:136 either miss valid functions or hit undefined behavior when the C string is compared to std::string. The name needs to be read from the integrated attribute itself, not from dwarf_diename() on the raw DIE.

**Comment by jordalgo** [comment]
> so maybe a wrapper could make it safer

That works for me. That being said if it's a pretty small edge case then I don't care as much (AI is still not good at knowing when things will actually happen in practice).

File: tests/testprogs/CMakeLists.txt (hunk 9 above)
Comment (MarekPazur) [reply tests/testprogs/CMakeLists.txt:112]: This is not needed at all. `.gnu_debuglink` is just an ELF section that includes a path string pointing to the `.debug` file's location, which we don't actually want here, since we're testing whether it can be found using the `--debuginfo` option. The `--add-gnu-debuglink` flag is useful if you want to hardcode the path so the DWARF aware tools can find it faster.

Comment (jordalgo) [reply tests/testprogs/CMakeLists.txt:112]: Ah ok. Sorry for the noise.

File: src/dwarf_parser.h [hunk 10]
```diff
@@ -87,15 +116,15 @@ class Dwarf {
...
-      std::string_view cu_comp_dir);
+  static std::optional<std::filesystem::path> get_cu_src_path(Dwarf_Die *cudie);
 
   Dwfl *dwfl = nullptr;
   Dwfl_Callbacks callbacks;
 
   BPFtrace *bpftrace_;
   std::string file_path_;
+  std::string debuginfo_path_;
+  const char *debuginfo_path_cstr_;
```
Comment (viktormalik) [inline src/dwarf_parser.h:135]: Please add a comment why we need both these fields (to avoid someone unifying them in future).

File: src/dwarf_parser.cpp [hunk 11]
```diff
@@ -27,24 +28,49 @@ struct FuncInfo {
...
+  callbacks.debuginfo_path = const_cast<char **>(&debuginfo_path_cstr_);
   dwfl = dwfl_begin(&callbacks);
   dwfl_report_offline(dwfl, file_path.c_str(), file_path.c_str(), -1);
   dwfl_report_end(dwfl, nullptr, nullptr);
 }
 
+static bool debug_alt_link_missing(::Dwarf *dwarf)
+{
+  const char *alt_name;
+  const void *build_id;
```
Comment (viktormalik) [inline src/dwarf_parser.cpp]: Should we initialize these to NULL?
```suggestion
  const char *alt_name = nullptr;
  const void *build_id = nullptr;
```

File: src/dwarf_parser.cpp [hunk 12]
```diff
@@ -54,11 +80,69 @@ Dwarf::~Dwarf()
...
+  return true;
+}
+
+static std::string get_die_name(Dwarf_Die *die)
+{
+  // If the DWARF section containing the DIE name string is malformed or
+  // missing, dwarf_diename() will return nullptr, so we wrap it to prevent
+  // undefined behaviour.
+  const char *name = dwarf_diename(die);
+  return name ? name : "";
```
Comment (viktormalik) [inline src/dwarf_parser.cpp]: Nit, you could use
```suggestion
  return name ?: "";
```

File: tests/testprogs/CMakeLists.txt [hunk 13]
```diff
@@ -96,6 +96,51 @@ add_custom_command(
...
+  COMMAND ${LLVM_OBJCOPY} --strip-debug $<TARGET_FILE:uprobe_test> ${CMAKE_CURRENT_BINARY_DIR}/uprobe_test-stripped
+  DEPENDS uprobe_test
+)
+list(APPEND testprogtargets
+  ${CMAKE_CURRENT_BINARY_DIR}/uprobe_test-stripped
+  ${CMAKE_CURRENT_BINARY_DIR}/debug/uprobe_test-stripped.debug)
+
+# Build a binary with split DWARF (.dwo files) to verify parsing of the split debug info format.
+add_custom_command(
+  OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/debug/dwo/uprobe_test-split
```
Comment (viktormalik) [inline tests/testprogs/CMakeLists.txt]: Shouldn't we also make the `.dwo` file an output of this command?

File: src/dwarf_parser.cpp (hunk 11 above)
Comment (MarekPazur) [reply src/dwarf_parser.cpp]: Not needed, as the [dwelf function](https://git.sr.ht/~sourceware/elfutils/tree/elfutils-0.194/item/libdwelf/dwelf_dwarf_gnu_debugaltlink.c) uses these variables only to assign a `char*` string to them. But, I’ll do it anyway to make it safe.
